migration to 14 and add repo reference

matthieusaison opened the pull request.

Commit:
migration to 14 and add repo reference

Commit:
migrate module_info_push

Add param to exclude custom module from the list of module info
Apply precommit on all the repo

Commit:
add setup/ for precommit

sebastienbeau commented on module_info_push/models/ir_module_module.py:
```diff
@@ -55,3 +70,17 @@ def push_installed_module_info(self):
         return data
+
+    def _compute_is_custom_module(self):
+        for record in self:
```
You can make it simplier.
```
custom_path = self.env["ir.config_parameter"].sudo().get_param("module_info.custom_path")
for record in self:
    module_path = get_module_path(record.name)
    record.is_custom_module = module_path and custom_path in module_path
```

sebastienbeau commented on module_info_partner/models/module_information.py:
```diff
@@ -1,12 +1,25 @@
     _description = "Module Information and availability"
 
-    # TODO add unique constraint
+    # TODO add unique constraint, see proposal line 69
```
remove comment

sebastienbeau commented on module_info_partner/models/module_information.py:
```diff
@@ -53,6 +69,15 @@ class ModuleInformation(models.Model):
         "this one and could replace it.",
     )
 
+    # TO VALIDATE: (cf anotation add unique constraint)
```
ok, you can remove the comment

Commit:
Simplify custom module compute

add website in manifest

matthieusaison committed:
Simplify custom module compute

add website in manifest

Commit:
Merge branch '14.0-mig-partner-module-information' of https://github.com/akretion/partner-module-information into 14.0-mig-partner-module-information

Commit:
add test for push info

always sent is_custom in push infos

Commit:
import repo and modules data from yaml

add cron to do it weekly
add view to shox host repository informations

Commit:
add test on import module from yaml

Commit:
add module info pull request

    fix test for info import
    rename field organisation
    delete field ModuleInformation.description_html

Commit:
Fix install of module_info_push

sebastienbeau commented on module_info_import/__manifest__.py:
```diff
@@ -1,12 +1,12 @@
     "name": "Module Info Import",
-    "description": "Get information about modules from github connector odoo",
-    "version": "12.0.1.0.0",
+    "summary": "Get information about modules from github connector odoo",
```
Update the description

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,43 +1,40 @@
 
+    @api.model
+    def get_module_info(self, version):
+        # make it multi version 14.0 /15.0 /16.0
```
remove comment

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,43 +1,40 @@
+        )
+        response = requests.get(url)
+        modules_list = yaml.safe_load(response.text)
+        return modules_list
```
direct return to avoid to define a variable
```suggestion
        return yaml.safe_load(response.text)
```

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -49,59 +46,46 @@ def _get_module_information_vals_from_github_stat(self, module_info):
-                                }
-                            )
+        # create Host orga & repo:
+        host_vals = {"organization": orga, "name": repo}
+        host_id = self.env["host.repository"].search([("name", "=", repo)], limit=1)
+
+        if host_id and host_id.organization != orga:
+            self.env["host.repository"].create(host_vals)
+        elif not host_id:
+            host_id = self.env["host.repository"].create(host_vals)
+
+        # check version
```
code can  be simplier.
improve naming 
repo => repo_name
"host_vals" => "repo_vals"
"host_id" => repo (never add _id if it's an object)
search on "name" and "organization" (right now you only search on name, this can generate error)

sebastienbeau replied:
repo = self._get_or_create_repo(self, repo_name, orga_name)

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -49,59 +46,46 @@ def _get_module_information_vals_from_github_stat(self, module_info):
+        }
+
+        module = self.search([("technical_name", "=", module)])
+        if module:
```
TODO we should simplify this code.

First we can introduce a method on module.module

```
def _should_update_module(self, version, orga):
    self.ensure_one()
    orga_priority = {"oca": 100, "akretion": 50}
    return (
        max(self.mapped("available_version_ids.name")) <= version
        and orga_priority.get(module.repo_id.orga, 0) <= orga_priority.get(orga, 0) 
)


```

sebastienbeau replied:
```

def _add_available_version(self, version):
    if version not in self.mapped("available_version_ids.name"):
          self.env["module.version"].create({
              "module_id": self.id,
              "version_id": self._get_version(),
              "state": "done"
               })

def _update_module(self, info):
     ....


module = self.search([("name", "=", module_name), ("partner_id", "=", False)])
if module:
    if module._should_update_module(version, orga):
        # to update description, author, ....
        module._update_module(module_info)

    module._add_available_version(version)
else:
    # create module


```

sebastienbeau commented on module_info_partner/models/module_information.py:
```diff
@@ -53,6 +67,14 @@ class ModuleInformation(models.Model):
+    _sql_constraints = [
+        (
+            "uniq_technical_name",
+            "unique(technical_name)",
```
"unique(technical_name, partner_id)"

sebastienbeau commented on module_info_partner/models/module_information.py:
```diff
@@ -1,20 +1,34 @@
         help="Edit this field to store complementary information about the " "module"
     )
     authors = fields.Char(readonly=True)
+    host_repository_id = fields.Many2one(
```
rename host_repository_id => repo_id
add the field partner_id for custom

matthieusaison replied:
The partner storage for custom is done in module.partner: 
https://github.com/akretion/partner-module-information/blob/14.0-mig-partner-module-information/module_info_partner/models/module_partner.py#L8

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,43 +1,40 @@
 
 
 ERROR_MESSAGE = _("There is an issue with module information synchronization")
+VERSIONS = ["14.0", "15.0", "16.0"]
```
VERSIONS = ["14.0", "15.0", "16.0"]
SELECTION_VERSIONS = [(v, v) for v in VERSIONS]

Commit:
improve info import

Commit:
add modules ids in PR

Rename class host.repository in module.repo
add modules_ids to PR object from diff url
change odoo versions selection to text field
adjust migrations status

Commit:
add test

add fields reviewer_count on pull.resquest

Commit:
some fix

Commit:
Add all PR state from github api

rewrite update and create pr

sebastienbeau commented on module_info_partner/models/module_information.py:
```diff
@@ -1,20 +1,22 @@
+    repo_id = fields.Many2one("module.repo", index=True, string="Host Repository")
+    partner_id = fields.Many2one(
+        "res.partner", index=True, string="Partner's Custom Module"
+    )
```
module_partner_ids  = fields.one2many("module.partner", "module_id")

sebastienbeau replied:
puis sur les PR on pourrait filtrer avec

```
<field name="module_ids" string="Customer" filter_domain="[('module_ids.module_partner_ids.partner_id', 'ilike', self)]"/>

``

Commit:
Use api github search rather pulls

move import_pr from model pull.request to model module.repo
handle the date_last_updated to ask only the new PR's events

Commit:
Add smart button on PR form view

Commit:
Add smart button on module repo view

add groupby on module repo view
add smart button on module information

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
 
 
 ERROR_MESSAGE = _("There is an issue with module information synchronization")
+VERSIONS = ["14.0", "15.0", "16.0"]
```
can you replace it by the "odoo.version" object ?

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
-            raise UserError(ERROR_MESSAGE)
-        return data
+    def synchronize_module(self):
+        for version in VERSIONS:
```
use odoo.version

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
-            "description_html": module_info["description_rst_html"],
-            "authors": module_info["author_ids_description"],
+    def _update_or_create_modules(self, version, orga_name, repo_name, module):
+        odoo_version_dct = {
```
this can be remove if we pass the "odoo.version" as object in version

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
+        vals = {
+            "available_version_ids": [(4, version_id, 0)],
+            "repo_id": repo.id,
+            "technical_name": module,
```
TODO we should add the module description

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
+        repo = self._get_or_create_repo(orga_name, repo_name)
+        version_id = odoo_version_dct.get(version, "odoo_version_unknown")
+        vals = {
+            "available_version_ids": [(4, version_id, 0)],
```
Note needed as we add the version after
```suggestion
```

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
+        repo = self.env["module.repo"].search(
+            [("name", "=", repo_name), ("organization", "=", orga_name)], limit=1
         )
+        if repo and repo.organization != orga_name:
```
simplify this code
Testing "repo.organization != orga_name" is useless as we search on the organization just before

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
+    def _get_or_create_repo(self, orga_name, repo_name):
+        repo_vals = {"organization": orga_name, "name": repo_name}
+        repo = self.env["module.repo"].search(
+            [("name", "=", repo_name), ("organization", "=", orga_name)], limit=1
```
limit=1 is not needed, just add an sql constraint

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -1,107 +1,92 @@
-            module = self.search(
-                [("technical_name", "=", module_info["technical_name"])]
+    @api.model
+    def _add_available_version(self, version):
+        if version not in self.mapped("available_version_ids.version"):
+            self.env["module.version"].create(
+                {
+                    "module_id": self.id,
+                    "version_id": self._get_version(),
+                    "state": "done",
+                }
```
Be careful version can existing from an existing PR migration

- note version should be an odoo.version object

```suggestion
    def _add_available_version(self, version):
        module_version = self.module_version_ids.filtered(lambda s: s.version_id == version)
        if module_version and module_version.state != "done":
            # In case that a version have been creating from a pending PR
            module_version.state = "done"
        else:
            self.env["module.version"].create(
                {
                    "module_id": self.id,
                    "version_id": version.id,
                    "state": "done",
                }
```

sebastienbeau commented on module_info_import/tests/test_module_info_partner.py:
```diff
@@ -0,0 +1,68 @@
+from odoo.tests import TransactionCase
+
+
+@patch("odoo.addons.module_info_import.models.module_information.VERSIONS", ["14.0"])
```
use odoo.version

sebastienbeau commented on module_info_partner/models/module_repo.py:
```diff
@@ -0,0 +1,22 @@
+    module_ids = fields.One2many(
+        "module.information", "repo_id", string="Module information"
+    )
+    module_ids_nbr = fields.Integer(
```
```suggestion
    module_nbr = fields.Integer(
```

sebastienbeau commented on module_info_pull_request/models/module_informations.py:
```diff
@@ -0,0 +1,25 @@
+    _inherit = "module.information"
+
+    pr_ids = fields.Many2many("pull.request", string="Active Pull Request")
+    pr_ids_nbr = fields.Integer(compute="_compute_pr_ids_nbr", string="# of Modules")
```
```suggestion
    pr_nbr = fields.Integer(compute="_compute_pr_ids_nbr", string="# of Modules")
```

sebastienbeau commented on module_info_pull_request/models/module_informations.py:
```diff
@@ -0,0 +1,25 @@
+            "module_info_pull_request.pull_request_action"
+        )
+        if len(prs) > 0:
+            action["domain"] = [("id", "in", prs.ids)]
```
passing all ids is not a good practice (the list can be very long)

Moreover it can be done without any python code.
Add a filter in view of the pull.request on repo_id and activate it with the context

```
<button name="%(pull_request_action)d" type="action" context="{'search_default_repo_id': active_id}"/>
```

example : https://github.com/odoo/odoo/blob/264af51db8a11177e2792c1426640f5ae31332f9/addons/maintenance/views/maintenance_views.xml#L907

matthieusaison replied:
Ok, it's prettier but how can I pass the PR ids to display on pr view?

matthieusaison replied:
It's work, when I use a search view.

sebastienbeau commented on module_info_pull_request/models/module_repo.py:
```diff
@@ -0,0 +1,90 @@
+
+    # updated_at = fields.Datetime(string="Last Update")
+    date_last_updated = fields.Datetime(
+        string="Last Update date", default="2023-01-01"
```
hum, default should be empty.
And if the date is empty (first run) then we should only download open PR

Commit:
Add search on project's partner on module_information

Commit:
Changes after review

Replace smart button associated method by usage of active_id
Use odoo.version rather VERSIONS list
Fix test
Rewrite module creation/update
Add constraint on module.repo uniq:(name,organization)
Rename some field

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -13,95 +14,87 @@
+        module_version = self.module_version_ids.filtered(
+            lambda s: s.version_id == version
+        )
+        if module_version and module_version.state != "done":
+            # In case that a version have been creating from a pending PR
+            module_version.state = "done"
```
```suggestion
        if module_version:
            if module_version.state != "done":
                # In case that a version have been creating from a pending PR
                module_version.state = "done"
```

Commit:
Change calling of api github to /pulls

handle update with date_last_updated
for new repo, call all opened pr
for update call all state and stop pagination when date_last_updated reached

Commit:
Enhance module_repo smart button by using context

Commit:
Add Kanban view

Commit:
Add PR stats to Kanban partner project

Rename crons in order to regroup then under prefix
Handle custom partner module (no community) with add of partner_id on create

Commit:
rename module_partner_migration to module_info_migration

Commit:
Improve views

Move res_partner code tomodule  pull request
Change smart button of PR by a compute domain (fix bug active_id not update in nested smart button)
Add filter on PR view
Add the number of module on kanban partner project view

Commit:
Improve views

switch some field to readonly
use widget for m2m fields
add groupby filter

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -13,95 +14,84 @@
+            module._add_available_version(version)
+        else:
+            vals.update({"available_version_ids": [(4, version.id, 0)]})
+            _logger.info(">>>> CREATE MODULE : %s", vals)
```
remove logger

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -13,95 +14,84 @@
         )
+        if not repo:
+            repo = self.env["module.repo"].create(repo_vals)
+        return repo
```
It's weird to define the variable  `repo_vals` even if we do not need it.
Maybe something like that will be more readable

```suggestion
    def _get_or_create_repo(self, orga_name, repo_name):
        repo = self.env["module.repo"].search(
            [("name", "=", repo_name), ("organization", "=", orga_name)]
        )
        if repo:
            return repo
        else:
            return self.env["module.repo"].create({
                "organization": orga_name,
                "name": repo_name
                })
```

sebastienbeau commented on module_info_import/models/module_information.py:
```diff
@@ -13,95 +14,84 @@
-                [("technical_name", "=", module_info["technical_name"])]
+    @api.model
+    def _add_available_version(self, version):
+        # import pdb; pdb.set_trace()
```
remove comment

sebastienbeau commented on module_info_migration/models/module_partner.py:
```diff
@@ -0,0 +1,43 @@
+from odoo import api, fields, models
```
Danger, you have rename the module "module_partner_migration" to "partner_info_migration", you should have use "git mv". The file have been added without deleting the previous one.
I am going to fix it

Commit:
add timesheet on pr view

Commit:
Fix after review

sebastienbeau commented on module_info_pull_request/models/pull_request.py:
```diff
@@ -0,0 +1,168 @@
+            line.task_id = False
+
+    @api.depends("reviewer_ids")
+    def _compute_reviewer(self):
```
```suggestion
    def _compute_reviewer_count(self):
```

sebastienbeau committed:
module_info_partner: add icon, fix deleting, fix import of custom module

Commit:
Add moduletimesheet

Add the timesheet to the pr view
Auto compute the timesheet description

sebastienbeau committed:
copier update

Commit:
Add new vals from module tracker

Rename technical_name to name

Commit:
Move module_nbr to module_info_partner

sebastienbeau committed:
module_info_partner, module_info_import: better naming and fix push for customer

sebastienbeau committed:
module_info_partner: fix access right issue

Commit:
Rename module info timesheet

sebastienbeau committed:
fixup! module_info_partner: fix access right issue

Commit:
Add group by

Commit:
Add stored pr_nbr

Commit:
Add list of modue user on module information

Add field github user on res user in order to filter on my PR
Add filter onview pull request to see
	my PR,
	my review,
	available review

Commit:
update readme

bguillot committed:
fix dont crash if readme and authors are empty in case of old modules

bguillot committed:
change field description to html type and adapt with the module file

bguillot committed:
Merge pull request #9 from akretion/14-fix-old-modules

fix dont crash if readme and authors are empty in case of old modules

florian-dacosta committed:
[FIX] Avoid crash when a module exists but has no available version
This can happen if the module was first created by a partner instance push.

bguillot committed:
Merge pull request #10 from akretion/14-fix-github-update-crash

[FIX] Avoid crash when a module exists but has no available version

florian-dacosta committed:
[FIX] Avoid crash of the sync cron when the version is not present in the github yaml files

florian-dacosta committed:
Merge pull request #11 from akretion/14.0-fix-cron-old-version

[FIX] Avoid crash of the sync cron when the version is not present in…

florian-dacosta committed:
[IMP] Manage pending module version when synchronizing the PR
This allows to full the in progress versions on the module and is quite usefull for a migration process

florian-dacosta committed:
Merge pull request #12 from akretion/14-manage-pending-module-version-from-pr

[IMP] Manage pending module version when synchronizing the PR

hparfr commented:
hello, @sebastienbeau @matthieusaison 

what is the status of this PR ?

hparfr commented on module_info_partner/models/module_partner.py:
```diff
@@ -30,17 +49,26 @@ def _prepare_partner_module_vals(self, partner, version, module, module_info):
+            module = module_info_obj.create(
+                self._prepare_module_info_vals(module_info, partner)
+            )
+        version = self.env["odoo.version"].search([("name", "=", version_num)])
```
this field should be protected if there is no version_num already registered (like v17) , it fails at the SQL INSERT statment.

So, either improve the error message, or allow to create new version here and take care to handle  how master version

hparfr commented:
How entreprise edition should be handled ?

I'm running 14.0+e , so should  +e` be stripped on the client (before sending) to looks like a normal v14 ?

sebastienbeau commented:
I merge the pull request. Additionnal improvement will be done in a separated PR
